### DISCUSSION TITLE
.chezmoiexternal + chezmoi add == inconsistent state?

### DISCUSSION THREAD

**Discussion by ToxicFrog** [twpayne/chezmoi#2956] [Q&A]
I have a `~/.zsh/` directory. It holds all my rc scripts for zsh, as well a git repo holding the ZGen package manager. I am trying to migrate this setup to chezmoi.

So, I create `.chezmoiexternal.toml`:
```
[".zsh/zgen"]
  type = "git-repo"
  url  = "https://github.com/tarjoilija/zgen.git"
```
and use it for a bit and everything is fine.

But then I edit my zsh configuration, and I don't want to manually add every file in `.zsh` that I edited, so I just `chezmoi add ~/.zsh`. And that succeeds, or at least appears to! But then I `chezmoi status` and:
```
$ chezmoi status    
chezmoi: .zsh/zgen: inconsistent state (.local/share/chezmoi/dot_zsh/zgen, https://github.com/tarjoilija/zgen.git defined in .local/share/chezmoi/.chezmoiexternal.toml)
```

Looking at my source dir reveals that `chezmoi add` has created a `dot_zsh/zgen` directory. Apparently `chezmoi add` was not able to realize that `.zsh/zgen` was managed by `.chezmoiexternal` and tried to add it normally, thus the inconsistency.

I can work around this by adding it to `.chezmoiignore`:
```
.zsh/zgen/**
```
but that produces a big ball of warning spam.

I've also tried using git submodules with the `external_` prefix [as documented here](https://www.chezmoi.io/user-guide/include-files-from-elsewhere/#include-a-subdirectory-from-a-git-repository), which work a *lot* better overall, but have a different problem: even with the `external_` prefix, the executable bit isn't preserved for any files in that submodule.

Ideally, I think, the behaviour I would like is:
- setting up cm on a new machine clones the git repos in .chezmoiexternal into the appropriate directory
- edits I make in those directories show up as diffs
- `cm add` copies those edits into the cm repo, and `cm apply` discards them

`.chezmoiexternal` seems to be the preferred approach for this, but I can't figure out how to get that behaviour. Submodules are a lot closer, apart from the executable bit getting dropped.

Anyone know how to do this?

**Comment by nandalopes** [comment]
The possible workaround is to delete `dot_zsh/zgen` from source state after running `chezmoi add`.

**Comment by Lockszmith-GH** [comment]
Maybe add `dot_zsh/zgen` to both `.gitignore` and [`.chezmoiignore`](https://www.chezmoi.io/reference/special-files-and-directories/chezmoiignore/) to prevent it from being added?

Not sure as I've never tested it, but it seems that would keep you from chezmoi wanting to deal with that path.

**Comment by twpayne** [comment]
Thanks for reporting this!

Fundamentally, I think you've hit #1574 which will require a refactor of chezmoi's internals to fix. This is planned (but not promised) for [chezmoi v3](https://github.com/twpayne/chezmoi/discussions/2673).

> Submodules are a lot closer, apart from the executable bit getting dropped.

You created a fantastic bug report in #2964 and this should now be fixed with #2966 and released in [`v2.33.4`](https://github.com/twpayne/chezmoi/releases/tag/v2.33.4)
